--- a/drivers/net/phy/Kconfig
+++ b/drivers/net/phy/Kconfig
@@ -318,6 +318,14 @@ config RTL8367_PHY
 	tristate "Driver for the Realtek RTL8367R/M switches"
 	select SWCONFIG
 
+config RTL8367B_PHY
+	tristate "Driver for the Realtek RTL8367R [B/-VB] switch"
+	select SWCONFIG
+
+config RTL8367_MDIO
+        tristate "RTL8367R MDIO interface"
+        select SWCONFIG
+
 endif # PHYLIB
 
 config MICREL_KS8995MA
--- a/drivers/net/phy/Makefile
+++ b/drivers/net/phy/Makefile
@@ -32,6 +32,9 @@ obj-$(CONFIG_RTL8366_SMI)	+= rtl8366_smi
 obj-$(CONFIG_RTL8366S_PHY)	+= rtl8366s.o
 obj-$(CONFIG_RTL8366RB_PHY)	+= rtl8366rb.o
 obj-$(CONFIG_RTL8367_PHY)	+= rtl8367.o
+obj-$(CONFIG_RTL8367_MDIO)      += mii_mgr.o
+obj-$(CONFIG_RTL8367_MDIO)      += rtl8367_mdio.o
+obj-$(CONFIG_RTL8367B_PHY)	+= rtl8367b.o
 obj-$(CONFIG_LSI_ET1011C_PHY)	+= et1011c.o
 obj-$(CONFIG_FIXED_PHY)		+= fixed_phy.o
 obj-$(CONFIG_MDIO_BITBANG)	+= mdio-bitbang.o
